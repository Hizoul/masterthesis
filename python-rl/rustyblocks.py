#!/usr/bin/env python3
import math
import sys, ctypes
from os.path import abspath, dirname, realpath
from ctypes import c_char_p, c_uint32, c_uint8, c_int8, Structure, POINTER,c_void_p,c_int64,c_int16,c_double,c_uint64,c_char_p
import numpy as np
from numpy.ctypeslib import ndpointer
from PIL import Image


parent_folder = dirname(realpath(__file__)) + "/"
class Tuple(Structure):
    _fields_ = [("placed", c_uint8),("reward", c_double),("done", c_uint8),("winner", c_uint8)]
    def __str__(self):
        return "({},{},{},{})".format(self.placed, self.reward,self.done,self.winner)

class GameFieldS(Structure):
    pass

# libpath = "../target/debug/"
libpath = "../target/release/"
libname = "rustyblocks"
prefix = {'win32': ''}.get(sys.platform, 'lib')
extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
absolute_lib_path = abspath(parent_folder + libpath + prefix + libname + extension)
print("Loading rustyblocks from ", absolute_lib_path)
rustLib = ctypes.cdll.LoadLibrary(absolute_lib_path)

rustLib.field_new.restype = POINTER(GameFieldS)
rustLib.field_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_new.restype = POINTER(GameFieldS)
rustLib.field_clone.argtypes = (POINTER(GameFieldS), )
rustLib.field_clone.restype = POINTER(GameFieldS)
rustLib.hex_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.hex_do_action_with_answer.restype = Tuple
rustLib.hex_to_array.restype = ndpointer(dtype=ctypes.c_uint8, shape=(121,))
rustLib.free_hex_array.argtypes = (ndpointer(dtype=ctypes.c_uint8, shape=(121,)), )
rustLib.hex_reset.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.restype = c_void_p
rustLib.field_to_json_log.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_json_log.restype = c_void_p
rustLib.field_restore_log.argtypes = (POINTER(GameFieldS), c_char_p)
rustLib.field_to_array.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_to_array.restype = ndpointer(dtype=c_int8, shape=(420,))
rustLib.free_field_array.argtypes = (ndpointer(dtype=c_int8, shape=(420,)), )
rustLib.field_get_score.argtypes = (POINTER(GameFieldS),)
rustLib.field_get_score.restype = ndpointer(dtype=c_int16, shape=(2,))
rustLib.free_score_array.argtypes = (ndpointer(dtype=c_int16, shape=(2,)), )
rustLib.field_do_action.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_do_action.restype = c_uint8
rustLib.field_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.field_do_action_with_answer.restype = Tuple
rustLib.field_do_action_with_answer_heu.argtypes = (POINTER(GameFieldS), c_double,c_double,c_double,c_double, c_uint8)
rustLib.field_do_action_with_answer_heu.restype = Tuple
rustLib.field_do_action_self_play.argtypes = (POINTER(GameFieldS), c_uint8, c_int8)
rustLib.field_do_action_self_play.restype = Tuple
rustLib.field_is_game_over.argtypes = (POINTER(GameFieldS), )
rustLib.field_is_game_over.restype = c_uint8
rustLib.field_get_reward.argtypes = (POINTER(GameFieldS), )
rustLib.field_get_reward.restype = c_double
rustLib.field_counter_action.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action.restype = c_int64
rustLib.field_counter_action_index.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action_index.restype = c_uint8
rustLib.field_reset.argtypes = (POINTER(GameFieldS), )
rustLib.free_string.argtypes = (c_void_p, )
rustLib.eval_heuristic_weights.argtypes = (c_double,c_double,c_double,c_double)
rustLib.eval_heuristic_weights.restype = c_uint64

def field_to_string(field):
  ptr = rustLib.field_to_string(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)
def field_to_log(field):
  ptr = rustLib.field_to_json_log(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)

def field_to_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy().reshape((42,10))
  rustLib.free_field_array(original_array)
  return result

def field_score(field):
  original_array = rustLib.field_get_score(field)
  result = original_array.copy()
  rustLib.free_score_array(original_array)
  return result

def field_to_cnn_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy()
  newres = np.zeros(shape=(32,32,3), dtype=np.uint8)
  for i in range(0, 200):
    y = math.floor(i / 10)
    x = i % 10
    current = 0
    val = 0
    if result[i] == 1 or result[i] == 2:
      current = 0
      if result[i] == 1:
        val = 128
      else:
        val = 255
    elif result[i] == 3 or result[i] == 4:
      current = 2
      if result[i] == 3:
        val = 128
      else:
        val = 255
    newres[y][x][current] = val
  for i in range(200, 210):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i] + 124
    newres[y][x][1] = val
  for i in range(210, 230):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    elif val > 0:
      val = val * 50
    newres[y][x][1] = val
  for i in range(230, 420):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    else:
      val = 255
    newres[y][x][1] = val
  rustLib.free_field_array(original_array)
  return newres

def hex_to_array(field):
  original_array = rustLib.hex_to_array(field)
  result = original_array.copy().reshape((11,11))
  rustLib.free_hex_array(original_array)
  return result

def field_string_to_array(field_string):
  field_array = np.zeros((20,10), dtype=np.uint8)
  row = 0
  col = 0
  for i in range(len(field_string)):
    entry = field_string[i]
    if entry == '\n':
      col = 0
      row += 1
    elif row < 20 and col < 10:
      field_array[row][col] = ord(entry) - 48
      col += 1
  return field_array

def save_img(data, name):
  Image.fromarray(data, 'RGB').save(name)