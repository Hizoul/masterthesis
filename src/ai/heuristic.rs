use crate::ai::{Bot};
use crate::game_logic::field::{GameField};
use crate::game_logic::field::rl::LearnHelper;

pub struct HeuristicBot {
  mode: u8
}

impl HeuristicBot {
  pub fn new(mode: u8) -> HeuristicBot {
    HeuristicBot {
      mode
    }
  }
}

impl Bot for HeuristicBot {
  fn make_play(&mut self, field: &GameField, _: bool) -> u8 {
    let mut cloned = field.clone();
    match self.mode {
      0 => cloned.get_best_heuristic_play_for_params(4.95238700733393,0.115980839099051,0.0603795891552745,19.0148438477953, true, 0),
      1 => cloned.get_random_heuristic_play(),
      _ => cloned.get_best_heuristic_play(true)
    }
  }
  fn get_name(&self) -> String {
    let start = match self.mode {
      0 => "Tuned",
      1 => "Random",
      _ => "User"
    };
    format!("{} Heuristic", start).to_owned()
  }
}