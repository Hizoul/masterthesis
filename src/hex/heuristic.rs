use super::{Coordinate,get_neighbors,coord_to_num,num_to_coord};

pub fn hex_heuristic(field: &Vec<Vec<u8>>, player: u8) -> usize {
  let field_size = field.len() - 1;
  let mut start_positions = Vec::new();
  let mut end_positions = Vec::new();
  for i in 0..field.len() {
    if player == 2 {
      start_positions.push((i, 0));
      end_positions.push((i, field_size));
    } else {
      start_positions.push((0, i));
      end_positions.push((field_size, i));
    }
  }
  let mut paths = Vec::new();
  for start in start_positions {
    for end in end_positions.iter() {
      paths.push(shortest_path(field, player, &start, end));
    }
  }
  if paths.len() > 0 {
    paths.sort_by(|a, b| a.len().partial_cmp(&b.len()).unwrap());
    for path in paths.iter() {
      for coord in path.iter() {
        if field[coord.1][coord.0] == 0 {
          return coord_to_num(field.len(), coord);
        }
      }
    }
  }
  // println!("COULDN*T FIND ANYTHING IN {:?}", paths);
  field.len() * 2
}

pub fn min_distance(dist: &mut Vec<usize>, visited: &mut Vec<bool>) -> usize {
    let mut min = std::usize::MAX;
    let mut min_index = 0;
    for i in 0..dist.len() {
      if visited[i] == false && dist[i] <= min {
        min = dist[i];
        min_index = i
      }
    }
    min_index 
}

pub fn shortest_path(field: &Vec<Vec<u8>>, player: u8, start: &Coordinate, end: &Coordinate) -> Vec<Coordinate> {
  let size = field.len() * field.len();
  let mut dist: Vec<usize> = Vec::with_capacity(size);
  let mut prev: Vec<usize> = Vec::with_capacity(size);
  let mut visited: Vec<bool> = Vec::with_capacity(size);
  for _ in 0..size {
    dist.push(std::usize::MAX);
    visited.push(false);
    prev.push(std::usize::MAX);
  }
  let mut already_has_block = false;
  for y in 0..field.len() {
    for x in 0..field.len() {
      if field[y][x] == player {
        already_has_block = true;
        dist[coord_to_num(field.len(), &(x,y))] = 1;
      }
    }
  }
  if !already_has_block {
    dist[coord_to_num(field.len(), start)] = 1;
  }
  let mut end_num = coord_to_num(field.len(), end);
  'SEARCH: for i in 0..size {
    let u = min_distance(&mut dist, &mut visited);
    visited[u] = true;
    if u == end_num {
      break 'SEARCH;
    }
    let coord = num_to_coord(field.len(), u);
    if coord.0 < field.len() && coord.1 < field.len() {
      let neighbors = get_neighbors(field.len(), coord.0, coord.1);
      for neighbor_option in neighbors.iter() {
        if neighbor_option.is_some() {
          let neighbor = neighbor_option.unwrap();
          if neighbor.0 < field.len() && neighbor.1 < field.len() {
            if field[neighbor.1][neighbor.0] == 0 {
              let v = coord_to_num(field.len(), &neighbor);
              if !visited[v] && dist[u] != std::usize::MAX && dist[u]+1 < dist[v] {
                dist[v] = dist[u] + 1;
                prev[v] = u;
              }
            }
          }
        }
      }
    }
  }
  let mut path = Vec::new();
  while end_num != std::usize::MAX {
    path.push(end_num);
    end_num = prev[end_num];
  }
  let mut coord_path = Vec::new();
  for n in path {
    coord_path.push(num_to_coord(field.len(), n));
  }
  return coord_path;
}

pub fn reshape(to_reshape: Vec<usize>) -> Vec<Vec<usize>> {
  let new_size = (to_reshape.len() as f64).sqrt() as usize;
  let mut new_top = Vec::with_capacity(new_size);
  for y in 0..new_size {
    let mut new_sub = Vec::with_capacity(new_size);
    for x in 0..new_size {
      new_sub.push(to_reshape[coord_to_num(new_size, &(x,y))]);
    }
    new_top.push(new_sub);
  }
  new_top
}