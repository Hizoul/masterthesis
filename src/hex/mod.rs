pub mod heuristic;
use rand::{Rng,thread_rng};

pub type Coordinate = (usize,usize);

fn num_to_coord(size: usize, action: usize) -> Coordinate {
  let y = action / size;
  let x = action % size;
  (x,y)
}

fn coord_to_num(size: usize, coord: &Coordinate) -> usize {
  coord.1 * size + coord.0
}

pub fn get_neighbors(field_size: usize, x: usize, y: usize) -> [Option<Coordinate>; 6] {
  [
    if x + 1 < field_size { Some((x + 1, y)) } else {None},
    if y + 1 < field_size { Some((x, y + 1)) } else {None},
    if x > 0 && y + 1 < field_size { Some((x - 1, y + 1)) } else {None},
    if x > 0 { Some((x - 1, y)) } else {None},
    if y > 0 { Some((x, y - 1)) } else {None},
    if x + 1 < field_size && y > 0 { Some((x + 1, y - 1)) } else {None}
  ]
}

#[derive(Clone)]
pub struct HexGame {
  pub field: Vec<Vec<u8>>,
  pub played_indices: Vec<usize>,
  pub is_first_players_turn: bool,
  pub game_over: bool,
  pub winner: u8,
  already_checked: Vec<Coordinate>,
  turn_counter: usize
}

impl HexGame {
  pub fn new(grid_size: usize) -> HexGame {
    let mut field = Vec::with_capacity(grid_size);
    for _ in 0..grid_size {
      let mut sub = Vec::with_capacity(grid_size);
      for _ in 0..grid_size {
        sub.push(0u8);
      }
      field.push(sub);
    }
    HexGame {
      field,
      is_first_players_turn: true,
      game_over: false,
      winner: 0,
      already_checked: Vec::with_capacity(grid_size * 3),
      turn_counter: 0,
      played_indices: Vec::with_capacity(grid_size*grid_size)
    }
  }
  pub fn place_point(&mut self, x: usize, y: usize) -> bool {
    // pie rule against first turn advantage. disabled so mcts has to outplay heuristic
    // if self.turn_counter == 1 {
    //   let placed_num = if self.is_first_players_turn {1} else {2};
    //   let opponent_num = if self.is_first_players_turn {2} else {1};
    //   if self.field[y][x] == opponent_num {
    //     self.turn_counter += 1;
    //     self.field[y][x] = placed_num;
    //     self.played_indices.push(self.global_to_local(coord_to_num(self.field.len(), &(x, y))));
    //     self.is_first_players_turn = !self.is_first_players_turn;
    //     self.check_game_over();
    //     return true;
    //   }
    // }
    if self.field[y][x] == 0 {
      self.turn_counter += 1;
      let placed_num = if self.is_first_players_turn {1} else {2};
      self.field[y][x] = placed_num;
      self.played_indices.push(self.global_to_local(coord_to_num(self.field.len(), &(x, y))));
      self.is_first_players_turn = !self.is_first_players_turn;
      self.check_game_over();
      return true;
    }
    false
  }
  pub fn do_action(&mut self, action: usize) -> bool {
    let size = self.field.len();
    let y = action / size;
    let x = action % size;
    self.place_point(x,y)
  }
  pub fn check_line(field: &Vec<Vec<u8>>, x: usize, y: usize, first_player: bool, already_checked: &mut Vec<Coordinate>) -> bool {
    let field_size = field.len() - 1;
    if x <= field_size && y <= field_size {
      let find_res = already_checked.iter().position(|pos| pos.0 == x && pos.1 == y);
      if find_res.is_none() {
        already_checked.push((x,y));
        let player_num = if first_player {1} else {2};
        if field[y][x] == player_num {
          if first_player && y == field_size || !first_player && x == field_size {
            return true;
          }
          let neighbors = get_neighbors(field.len(), x, y);
          for neighbor_option in neighbors.iter() {
            if neighbor_option.is_some() {
              let neighbor = neighbor_option.unwrap();
              if HexGame::check_line(field, neighbor.0, neighbor.1, first_player, already_checked) {
                return true;
              }
            }
          }
        }
      }
    }
    false
  }
  pub fn check_game_over(&mut self) {
    let field_size = self.field.len();
    let mut already_checked = &mut self.already_checked;
    for i in 0..(field_size*2) {
      let first_player = i < field_size;
      if (i % field_size) == 0 {already_checked.clear()}
      let x;
      let y;
      if first_player {
        x = i;
        y = 0;
      } else {
        x = 0;
        y = i % field_size;
      };
      let is_winning_line = HexGame::check_line(&self.field, x, y, first_player, &mut already_checked);
      if is_winning_line {
        self.game_over = true;
        self.winner = if first_player {1} else {2};
        return;
      }
    }
    self.game_over = true;
    let grid_size = self.field.len();
    for y in 0..grid_size {
      for x in 0..grid_size {
        if self.field[y][x] == 0 {
          self.game_over = false;
          return;
        }
      }
    }
  }
  pub fn get_random(&self) -> usize {
    if !self.game_over {
      let mut counter = 0;
      for _ in 0..50 {
        let x = thread_rng().gen_range(0, self.field.len() - 1) as usize;
        let y = thread_rng().gen_range(0, self.field.len() - 1) as usize;
        counter += 1;
        if self.field[y][x] == 0 {
          return coord_to_num(self.field.len(), &(x, y));
        }
        if counter > 50 {
        }
      }
    }
    for y in 0..self.field.len() {
      for x in 0..self.field.len() {
        if self.field[y][x] == 0 {
          return coord_to_num(self.field.len(), &(x,y));
        }
      }
    }
    0
  }
  pub fn random_place(&mut self) {
    if !self.game_over {
      let mut placed = false;
      while !placed {
        let x = thread_rng().gen_range(0, self.field.len() - 1) as usize;
        let y = thread_rng().gen_range(0, self.field.len() - 1) as usize;
        placed = self.place_point(x, y);
      }
    }
  }
  pub fn get_possible_plays(&self) -> usize {
    let mut play_amount = 0;
    let grid_size = self.field.len();
    for y in 0..grid_size {
      for x in 0..grid_size {
        if self.field[y][x] == 0 {
          play_amount += 1;
        }
      }
    }
    play_amount
  }
  pub fn local_to_global_action_index(&self, decision: usize) -> usize {
    let mut converted_decision = decision;
    let mut amount_free = 0;
    let grid_size = self.field.len();
    let max_action = (grid_size * grid_size)-1;
    'DONE_INCREASING: for y in 0..grid_size {
      for x in 0..grid_size {
        if self.field[y][x] != 0 {
          converted_decision += 1;
        } else {
          amount_free += 1;
        }
        // println!("AMOUNT FREE {} decision {} {}", amount_free, decision, converted_decision);
        if amount_free > decision || converted_decision >= max_action {
          break 'DONE_INCREASING;
        }
      }
    }
    // println!("converted {} {} {} {}", self.get_possible_plays(), decision, converted_decision, self);
    converted_decision
  }
  pub fn global_to_local(&self, action: usize) -> usize {
    let mut converted_decision = 0;
    for i in 0..action {
      let (x,y) = num_to_coord(self.field.len(), i);
      if self.field[y][x] == 0 {
        converted_decision += 1;
      }
    }
    converted_decision
  }
  pub fn to_1d_array(&self) -> Vec<u8> {
    let mut new_array = Vec::with_capacity(self.field.len()*self.field.len());
    for row in self.field.iter() {
      for val in row {
        new_array.push(*val);
      }
    }
    new_array
  }
  pub fn reset(&mut self) {
    self.is_first_players_turn = true;
    self.game_over = false;
    self.winner = 0;
    self.already_checked.clear();
    self.played_indices.clear();
    self.turn_counter = 0;
    let grid_size = self.field.len();
    for y in 0..grid_size {
      for x in 0..grid_size {
        self.field[y][x] = 0;
      }
    }
  }
}

impl std::fmt::Display for HexGame {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
      let grid_size = self.field.len();
      let mut res = String::new();
      for y in 0..grid_size {
        for _ in 0..y {
          res.push_str(" ");
        }
        for x in 0..grid_size {
          res.push_str(self.field[y][x].to_string().as_str())
        }
        res.push_str("\n");
      }
      write!(f, "{}", res)
    }
}

pub mod tournament;
pub mod players;
pub mod hex_eval;