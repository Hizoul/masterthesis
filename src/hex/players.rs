use crate::hex::tournament::{HexBot};
use crate::hex::heuristic::hex_heuristic;
use crate::hex::{HexGame};
use crate::mcts::rewrite::{MctsGame,MCTSBot};
use crate::mcts::tree::{Node,TreeBot};

impl MctsGame for HexGame {
  fn get_depth(&self) -> usize {self.turn_counter}
  fn get_action_amount(&self) -> usize {self.get_possible_plays()}
  fn get_current_player(&self) -> u8 {if self.is_first_players_turn {0} else {1}}
  fn do_action(&mut self, action: usize) {
    self.do_action(self.local_to_global_action_index(action));
  }
  fn get_action(&mut self, _: usize) -> usize {
    // to be changed for hex eval experiment
    self.global_to_local(self.get_random())
    // let answer = hex_heuristic(&self.field, if self.is_first_players_turn {1} else {2});
    // if answer == self.field.len() * 2 {
    //   self.global_to_local(self.get_random())
    // } else {
    //   self.global_to_local(answer)
    // }
    // self.global_to_local(hex_heuristic(&self.field, if self.is_first_players_turn {1} else {2}))
  }
  fn is_game_over(&self) -> bool {self.game_over}
  fn render_field(&self) -> Vec<u8> {self.to_1d_array()}
  fn get_winner(&self) -> i8 {
    if self.winner == 0 {3} else {(self.winner - 1) as i8}}
  fn box_clone(&self) -> Box<dyn MctsGame> {
    Box::new((*self).clone())
  }
}

pub struct HexHeuristicBot {}

impl HexHeuristicBot {
  pub fn new() -> HexHeuristicBot {
    HexHeuristicBot {
    }
  }
}

impl HexBot for HexHeuristicBot {
  fn make_play_hex(&mut self, field: &HexGame, is_second: bool) -> usize {
    let answer = hex_heuristic(&field.field, if is_second {2} else {1});
    if answer == field.field.len() * 2 {
      field.get_random()
    } else {
      answer
    }
  }
  fn get_name_hex(&self) -> String {
    "Heuristic".to_owned()
  }
}
pub struct HexRandomBot {
}

impl HexRandomBot {
  pub fn new() -> HexRandomBot {
    HexRandomBot {
    }
  }
}

impl HexBot for HexRandomBot {
  fn make_play_hex(&mut self, field: &HexGame, _: bool) -> usize {
    field.get_random()
  }
  fn get_name_hex(&self) -> String {
    "Random".to_owned()
  }
}

impl HexBot for MCTSBot {
  fn make_play_hex(&mut self, field: &HexGame, _: bool) -> usize {
    let parent = if field.turn_counter == 0 {
      self.get_root()
    } else {
      // println!("FINDING NEW PARENT ARC {:?}", field.played_indices);
      let mut parent_arc = self.get_root();
      let mut field_findhelper = HexGame::new(field.field.len());
      for i in field.played_indices.iter() {
        let i_usize = *i as usize;
        // println!("doing ACTION {} out of {} {} {}", field_findhelper.local_to_global_action_index(i_usize), field_findhelper.get_possible_plays(), i_usize, field_findhelper);
        field_findhelper.do_action(field_findhelper.local_to_global_action_index(i_usize));
        if !Node::child_is_parent(&parent_arc, i_usize) {
          parent_arc = Node::get_child(&parent_arc, i_usize);
        } else {
          parent_arc = Node::create_child(&parent_arc, i_usize, field_findhelper.get_possible_plays());
        }
      }
      parent_arc
    };
    let to_box = field.clone();
    let mut boxed = Box::new(to_box) as Box<dyn MctsGame>;
    let decision = self.mcts_algo(&mut boxed, &parent) as usize;
    // Node::print(&self.get_root());
    let converted_decision = field.local_to_global_action_index(decision);
    // println!("GOT MCTS DECISION {} converted to {}", decision, converted_decision);
    converted_decision
  }
  fn get_name_hex(&self) -> String {self.name.clone()}
}