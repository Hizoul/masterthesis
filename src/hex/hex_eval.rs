use crate::hex::tournament::{Tournament, HexBot, array_to_table};
use crate::hex::players::{HexHeuristicBot, HexRandomBot};
use crate::mcts::rewrite::MCTSBot;
use std::fs::{write};

const FILE_NAME: &str = "hex_final_tournament.json";

pub fn do_hex_eval() {
  let mut results: Vec<Vec<f64>> = Vec::with_capacity(3);
  let start = "heu";
  for _ in 0..3 {
    results.push(Vec::with_capacity(8));
  }
  for i in 2..12 {
    for mcts_version in 0..3 {
      let tournament_name = format!("{}_hex_mcts_{}_size_{}.json", start, mcts_version, i);
      let field_size = i;
      let mcts_action_size = field_size * field_size;
      let heu = HexHeuristicBot::new();
      let rand = HexRandomBot::new();
      let name = match mcts_version {
        1 => "MCTS-RAVE",
        2 => "MCTS-PoolRave",
        _ => "MCTS-UCB"
      };
      let mut m1 = MCTSBot::new_with_time(1000, mcts_action_size, name, mcts_action_size);
      if mcts_version > 0 {
        m1.config.use_rave = true;
      }
      if mcts_version > 1 {
        m1.config.use_pool_rave = true;
      }
      let mut t = Tournament::new(vec![
        Box::from(heu),
        Box::new(m1),
      ], 100, field_size);
      t.load(tournament_name.clone()).unwrap();
      t.run(Option::Some(tournament_name));
      t.make_ratings();
      println!("GOT WINS AGAINST {:?} {} {} {}", t.wins_against, mcts_version, t.wins_against[1][0], t.wins_against[1][0] / 100);
      let win_rate = t.wins_against[1][0] as f64;
      results[mcts_version].push(win_rate);
      // let player_names = t.player_names();
      // let mut elos = t.get_elos();
      // println!("{:?}", t.wins_against);
      // println!("{}", array_to_table(t.wins_against, player_names));
      // elos.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
      // println!("{:?}", elos);
    }
  }
  println!("got winrates {:?}", results);
  for mcts_version in 0..3 {
    let file_content = serde_json::to_string(&results[mcts_version]).unwrap();
    write(format!("{}_hex_{}.json", start, mcts_version), file_content).unwrap();
  }
}

pub fn do_hex_regular_tournament() {
  let field_size = 7;
  let mcts_action_size = field_size * field_size;
  let heu = HexHeuristicBot::new();
  let rand = HexRandomBot::new();
  let mut m1 = MCTSBot::new_with_time(1000, mcts_action_size, "MCTS-UCB", mcts_action_size);
  m1.config.simulation_amount = 5;
  let mut m2 = MCTSBot::new_with_time(1000, mcts_action_size, "MCTS-RAVE", mcts_action_size);
  m2.config.use_rave = true;
  let mut m3 = MCTSBot::new_with_time(1000, mcts_action_size, "MCTS-PoolRAVE", mcts_action_size);
  m3.config.use_rave = true;
  m3.config.use_pool_rave = true;
  let mut t = Tournament::new(vec![
    Box::new(m1),
    // Box::new(m2),
    // Box::new(m3)
    Box::from(heu),
    // Box::from(rand),
  ], 100, field_size);
  t.load(String::from(FILE_NAME)).unwrap();
  t.run(Option::Some(String::from(FILE_NAME)));
  t.make_ratings();
  let player_names = t.player_names();
  let mut elos = t.get_elos();
  println!("{:?}", t.wins_against);
  println!("{}", array_to_table(t.wins_against, player_names));
  elos.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
  println!("{:?}", elos);
}