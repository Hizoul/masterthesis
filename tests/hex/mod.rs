use rustyblocks::hex::{HexGame,get_neighbors};
use rustyblocks::hex::heuristic::hex_heuristic;
use rustyblocks::hex::hex_eval::{do_hex_eval,do_hex_regular_tournament};
use insta::{assert_snapshot};

#[test]
fn hex_game() {
  let mut game = HexGame::new(5);
  let mut game2 = HexGame::new(5);
  assert_snapshot!("initial_field", &game.to_string());
  for i in 0..5 {
    if i == 0 {
      game.place_point(0,i);
      game.place_point(2,2);
    } else {
      game.place_point(0,i);
      game.place_point(i,0);
    }
    assert_eq!(game.game_over, i == 4);
    assert_eq!(game.winner, if i == 4 {1} else {0});
    if i == 0 {
      game2.place_point(2,2);
      game2.place_point(i,0);
    } else {
      game2.place_point(0,i);
      game2.place_point(i,0);
    }
    assert_eq!(game2.game_over, i == 4);
    assert_eq!(game2.winner, if i == 4 {2} else {0});
  }
  assert_snapshot!("end", &game.to_string());
}

#[test]
fn hex_game_pie_rule() {
  // let mut game = HexGame::new(5);
  // assert_eq!(game.field[2][2], 0);
  // game.place_point(2,2);
  // assert_eq!(game.field[2][2], 1);
  // game.place_point(2,2);
  // assert_eq!(game.field[2][2], 2);
  // assert_eq!(game.place_point(2,2), false);
}

#[test]
fn hex_game_heuristic() {
  let mut game = HexGame::new(4);
  while !game.game_over {
    let action = hex_heuristic(&game.field, if game.is_first_players_turn {1}else{2});
    game.do_action(action);
  }
  assert_ne!(game.winner, 0);
  let mut wins_against_random = HexGame::new(11);
  while !wins_against_random.game_over {
    if wins_against_random.is_first_players_turn {
      let action = hex_heuristic(&wins_against_random.field, 1);
      wins_against_random.do_action(action);
    } else {
      wins_against_random.random_place();
    }
  }
  assert_eq!(wins_against_random.winner, 1);
}

#[test]
fn hex_neighbors_sanity_check() {
  let field_size: usize = 4;
  for x in 0..field_size {
    for y in 0..field_size {
      let neighbors = get_neighbors(field_size, x, y);
      for possible_neighbor in neighbors.iter() {
        if possible_neighbor.is_some() {
          let neighbor = possible_neighbor.unwrap();
          let neighbors_neighbors = get_neighbors(field_size, neighbor.0, neighbor.1);
          assert_eq!(neighbors_neighbors.iter().position(|pos| {
            if pos.is_some() {
              let unwrapped_pos = pos.unwrap();
              return unwrapped_pos.0 == x && unwrapped_pos.1 == y;
            }
            return false;
          }).is_some(), true);
        }
      }
    }
  }
}

#[test]
fn hex_game_do_action() {
  // do_hex_eval();
  // do_hex_regular_tournament();
}