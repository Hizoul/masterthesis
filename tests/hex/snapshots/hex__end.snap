---
source: tests/hex/mod.rs
expression: "&game.to_string()"
---
12222
 10000
  10200
   10000
    10000

